pkg/sync/sync_context.go

```diff
@@ -830,6 +850,83 @@ func (sc *syncContext) getSyncTasks() (_ syncTasks, successful bool) {
 	return tasks, successful
 }
 
+// processHookTasks applies additional logic to hook tasks.
+func (sc *syncContext) processHookTasks(tasks syncTasks) {
+	for _, task := range tasks {
+		// This is a safety check to ensure that we process only currently running hook tasks.
+		if !task.isHook() || !task.pending() {
+			continue
+		}
+		// Safety check to ensure that the target object is not nil.
+		if task.targetObj == nil {
+			continue
+		}
+		// Currently, we only process hook tasks where the target object is a Job.
+		if task.targetObj.GetKind() == "Job" {
+			sc.processJobHookTask(task)
+		}
+	}
+}
+
+// processJobHookTask processes a hook task where the target object is a Job and has defined ttlSecondsAfterFinished.
+// This addresses the issue where a Job with a ttlSecondsAfterFinished set to a low value gets deleted fast and the hook phase gets stuck.
+// For more info, see issue https://github.com/argoproj/argo-cd/issues/6880
+func (sc *syncContext) processJobHookTask(task *syncTask) {
+	hookFinalizer := "argoproj.io/hook-finalizer"
+
+	task.postprocess = func() error {
+		sc.log.V(1).Info("Processing hook task with a Job resource - attaching hook finalizer", "name", task.targetObj.GetName(), "namespace", task.targetObj.GetNamespace())
+
+		job, err := sc.clientset.BatchV1().Jobs(task.targetObj.GetNamespace()).Get(context.TODO(), task.targetObj.GetName(), metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		// Skip postprocessing if the Job does not have a ttlSecondsAfterFinished set.
+		if job.Spec.TTLSecondsAfterFinished == nil {
+			return nil
+		}
```

Comment:
What would be the downside(s) to always using a finalizer instead of only using one when this field is set?

Reply:
That is a good question, the logic would be simpler and more generic if all hooks have a finalizer, cleanup would also be simpler.